### PR TITLE
New API for confirming publishes

### PR DESCRIPTION
**Pull request by carlhoerberg**

### CODE REVIEW

**Review by baelter — APPROVED** [review]
lgmt, calls for new major version when released.

**Comment by dentarg** [comment]
What's the motivation? 

(Just so we don't forget, any update to the README needed?)

**Comment by carlhoerberg** [comment]
More discussion into why the `wait_for_confirms` doesn't always cut it: https://github.com/cloudamqp/amqp-client.cr/pull/38#issuecomment-1862370651

File: src/amqp-client/channel.cr [hunk 1]
```diff
@@ -256,24 +280,35 @@ class AMQP::Client
       if @confirm_mode
         @confirm_lock.synchronize do
           msgid = @confirm_id &+= 1_u64
-          @unconfirmed_publishes.push msgid
+          @unconfirmed_publishes.push({msgid, blk})
           msgid
         end
       else
+        raise ArgumentError.new("On confirm block received but channel is not in confirm mode") if blk
```
Comment (carlhoerberg) [inline src/amqp-client/channel.cr:287]: Right now we raise if a block is given to `basic_publish` but the channel isn't in confirm mode, should we instead automatically enable it?

File: src/amqp-client/channel.cr [hunk 2]
```diff
@@ -256,24 +280,35 @@ class AMQP::Client
...
 
     def basic_publish_confirm(msg, exchange, routing_key = "", mandatory = false, immediate = false, props properties = Properties.new) : Bool
       confirm_select
-      basic_publish(msg, exchange, routing_key, mandatory, immediate, properties)
-      wait_for_confirms
+      done = ::Channel(Nil).new
+      basic_publish(msg, exchange, routing_key, mandatory, immediate, properties) do
+        done.send nil
+      end
+      done.receive
```
Comment (carlhoerberg) [inline src/amqp-client/channel.cr:298]: This is probably very inefficient, creating a new `Channel` (on the heap) for each publish..

**Pull request merged**